File: lib/wus_control/programming_state.py
import board, busio, pwmio
from wus_control.state import State
from wus_control.hardware.AD9833 import AD9833
from wus_control.hardware.TUSS4470 import TUSS4470

class ProgrammingState(State):

    def __init__(self):
        # self._wave_gen = None
        self._wave_pwm = None
        self._burst_gen = None
        self._comm_port = None

    @property
    def name(self):
        return 'programming'

    def enter(self, controller):
        controller.rgb_led[0] = (255, 0, 0)
        State.enter(self, controller)
        if not self._burst_gen:
        # if not (self._wave_gen and self._burst_gen):    # When programming after advertising at startup
            self._startup_sequence(controller)
        elif controller.reprogram_setting:              # When programming single settings during runtime
            self._change_setting(controller)
        else:                                           # When entering programming state after readvertisement, no need to change settings
            pass

    def exit(self, controller):
        State.exit(self, controller)
        controller.reprogram_setting = None

    def update(self, controller):
        status = State.update(self, controller)
        if status == 'NO_CONNECT':
            controller.go_to_state('advertising')
        else:
            controller.go_to_state('idle')

    # Programs all settings
    def _startup_sequence(self, controller):
        controller.print("Startup: programming all settings...")
        st = controller.settings
        spi_port = busio.SPI(board.SCK, MOSI=board.MOSI, MISO=board.MISO)
        self._wave_gen = AD9833(st['cs_ad9833'])
        self._burst_gen = TUSS4470(st['cs_tuss4470'])

        # Program wave generator
        self._wave_pwm = pwmio.PWMOut(pin=st['pwm_wave_gen'], duty_cycle=2**15, frequency=int(2e6), variable_frequency=False)
        self._wave_gen.begin(spi_port)
        self._wave_gen.set_freq(st['frequency'])
        print(st['frequency'])
        self._wave_gen.set_type(1)
        self._wave_gen.send()

        # Program burst generator
        self._burst_gen.begin(spi_port)
        self._burst_gen.set(st['voltage'], st['current_mode'], st['io_mode'], st['pulse_count'])
        if st['regulated_mode']: self._burst_gen.enableRegulation()
        if st['pre_driver_mode']: self._burst_gen.enablePreDriver()
        controller.print('\tDone!')

    # If one setting gets changed during runtime, no need to reprogram all the settings
    def _change_setting(self, controller):
        setting = controller.reprogram_setting
        controller.print('Reprogramming ' + str(setting[0]) + ' to ' + str(setting[1]))
        controller.settings[setting[0]] = setting[1]
       
        if setting[0] =='frequency':
            self._wave_gen.set_freq(setting[1])
            self._wave_gen.send()
            # self._wave_pwm.frequency = setting[1]
        elif setting[0] == 'voltage':
            self._burst_gen.set_voltage(setting[1])
        elif setting[0] == 'pulse_count':
            self._burst_gen.set_pulse_count(setting[1])
        controller.print('\tDone!')

File: clientside/src/programming_state.cpp
#include "programming_state.h"

stateName ProgrammingState :: get_name(void) {
    return PROGRAMMING_STATE;
}

void ProgrammingState :: enter(StateController* ctrl) {
    // Serial.println("programming");
    ctrl->set_rgbLED(255, 255, 0); // yellow
    if (!this->initialized) {
        this->startup_sequence(ctrl);
    } else if (ctrl->settings->repgmKey != NO_CHG){
        this->change_setting(ctrl);
    }
}

void ProgrammingState :: exit(StateController* ctrl) {
    ctrl->settings->repgmKey = NO_CHG;
    ctrl->settings->repgmVal = 0;
    ctrl->devStatus = DEVICE_OK;
}

void ProgrammingState :: update(StateController* ctrl) {    
  if (this->initialized) {
    // Return to idle state if already connected
    ctrl->go_to_state(IdleState::getInstance());
  } else {
    // Go to advertising if not connected (usually only at program start)
    this->initialized = true;
    ctrl->go_to_state(AdvertisingState::getInstance());
  }
}

void ProgrammingState :: startup_sequence(StateController* ctrl) {
    #define refFreq 25000000UL

    BurstingState::initialize();
    
    ctrl->waveGen = new AD9833(ctrl->settings->cs_ad9833, refFreq);
    ctrl->burstGen = new TUSS4470(ctrl->settings->cs_tuss4470);

    ctrl->waveGen->Begin();
    ctrl->waveGen->DisableDAC(true);
    ctrl->waveGen->ApplySignal(SQUARE_WAVE, REG0, ctrl->settings->frequency);

    ctrl->burstGen->begin();
    ctrl->burstGen->set(ctrl->settings->voltage, ctrl->settings->current_mode, 
                  ctrl->settings->io_mode, ctrl->settings->pulse_count);
    if (ctrl->settings->regulated_mode) {ctrl->burstGen->enableRegulation();}
    if (ctrl->settings->pre_driver_mode) {ctrl->burstGen->enablePreDriver();}
    ctrl->burstGen->enableStandbyMode();
}

void ProgrammingState :: change_setting(StateController* ctrl) {
    uint32_t val = ctrl->settings->repgmVal;
    // Bit-cast to float if setting is a float by default
    float valFloat;
    memcpy(&valFloat, &val, 4);

    switch (ctrl->settings->repgmKey) {
        case BURSTPD_CHG:
            ctrl->settings->burst_pd = val;
            BurstingState::program_PWM();
            break;
            
        case BURSTDC_CHG:
            ctrl->settings->burst_dc = valFloat;
            BurstingState::program_PWM();
            break;

        case STIMPD_CHG:
            ctrl->settings->stim_pd = val;
            BurstingState::program_burst_timer();
            break;

        case STIMDC_CHG:
            ctrl->settings->stim_dc = valFloat;
            BurstingState::program_burst_timer();
            break;
        
        case FREQ_CHG:
            ctrl->settings->frequency = val;
            ctrl->waveGen->SetFrequency(REG0, val);
            break;
        
        case TOUT_CHG:
            ctrl->settings->timeout = val;
            BurstingState::program_timeout_timer();
            break;

        case VOLT_CHG: 
            ctrl->settings->voltage = val;
            ctrl->burstGen->setVoltage(val);
            break;

        case PULSECT_CHG: 
            ctrl->settings->pulse_count = val;
            ctrl->burstGen->setPulseCount(val);
            break;

        case DCSEQ_CHG:
            // dc_seq_len already changed in AdvertisingState::setting_rx_callback
            memcpy((void*) ctrl->settings->dc_seq_vals, reinterpret_cast<void*>(val), ctrl->settings->dc_seq_len);
            BurstingState::program_PWM();

        case NO_CHG:
            break;    
    }
}
